### ISSUE TITLE
Camera parameters

### ISSUE THREAD

**Issue by OrangeSodahub** [scannetpp/scannetpp#33]
Hi, I got some problems about the camera transformation, I transformed the boxes given in `segments_anno.json` to image plane, the script:

first I do the undistortion and downscale by 2,
```
from scipy.spatial.transform import Rotation
# get intrins
transform = json.load(open(os.path.join(scene_path, 'dslr', 'nerfstudio', f'{self.transform}.json'), 'r'))
fl_x, fl_y, cx, cy = transform['fl_x'], transform['fl_y'], transform['cx'], transform['cy']
intrin = (np.array([[fl_x, 0, cx, 0], [0, fl_y, cy, 0], [0, 0, 1, 0], [0, 0, 0, 1]]))
# get extrins
colmap = []
with open(os.path.join(raw_scene_path, 'dslr', 'colmap', 'images.txt'), 'rt') as f:
    for line in f:
        col = line.strip().split()
        if col[0] == '#' or len(col) > 10: continue
        colmap.append(col)
quaternions = np.array(colmap)[:, [2, 3, 4, 1]].astype(np.float32)
rots = Rotation.from_quat(quaternions).as_matrix()
trans = colmap[:, 5:8].astype(np.float32)
all_extrins = format_poses(rots, trans)
```
However I got these results:
![image](https://github.com/scannetpp/scannetpp/assets/54439582/11b276b4-087d-431d-b0d0-a57da5c58f95)

![image](https://github.com/scannetpp/scannetpp/assets/54439582/92e0a779-fa16-4300-b9a7-da078be9579c)

And I checked the camera positions, which are defined by `colmap[5:8]` above, and I got:
![image](https://github.com/scannetpp/scannetpp/assets/54439582/ffa00e14-f4b3-4b88-93a3-899fd3ecdc94)

where the green points show the camera positions, so wierd, do you have any advice?

**Comment by OrangeSodahub** [comment]
@cy94 @liu115 Hi, do you have any suggestion? Thanks!

**Comment by cy94** [comment]
Hi, did you follow the undistortion code in our repo?

**Comment by OrangeSodahub** [comment]
@cy94 Absolutely yes, where the problem might be ? The camera matrix after undistortion? It only changes `transforms.json`, not `colmap/images.txt`, so I get intrinsic matrix from the former, the extrinsic matrix from the latter, am I right? And I didn't change any code.

**Comment by OrangeSodahub** [comment]
And, forget about the camera matrix first, do you know why the camera positions are wrong? It comes from TX, TY, TZ in colmap/images.txt

**Comment by liu115** [comment]
Since I'm not seeing the code of how you plot the camera pose, I can think of a few things that might go wrong
1. colmap `images.txt` has **two** lines for each frame. The first line defines the camera poses, and the second one defines the feature points that are detected in the image.  It seems like you are reading them as camera poses in your code.
2. colmap qvec and tvec is the pose from world to camera to world and not the opposite.
3. Did you do additional axis alignment to the mesh and didn't apply it correctly to the camera poses?

**Comment by OrangeSodahub** [comment]
@liu115 Thanks for your reply,

1. I did skip the second line per frame by `len(col) > 10`;
2. What is "world to camera to world"?;
3. No additional operations.

**Comment by liu115** [comment]
my typo. The colmap poses are defined from world to camera. 

It's recommended to use the [reader from official colmap repo](https://github.com/scannetpp/scannetpp/blob/main/common/utils/colmap.py). Here is an example how to use it.

https://github.com/scannetpp/scannetpp/blob/f3c1af78b200e33a62afaad4d4d79a3cf6c2f151/common/render.py#L51


https://github.com/scannetpp/scannetpp/blob/f3c1af78b200e33a62afaad4d4d79a3cf6c2f151/common/render.py#L74

**Comment by OrangeSodahub** [comment]
@liu115 Thanks, I've checked your script and mine, the only difference is the qvec2rotmat function, however after I replace it with yours, it's still bad.

**Comment by OrangeSodahub** [comment]
As for the camera positions, you and I take the same operations, that's my plot-camera-pose script, you could check yourself:

```
import json
import numpy as np
from plyfile import PlyData
import matplotlib.pyplot as plt
from scipy.spatial.transform import Rotation


def read_ply(filename):
    """ read XYZ point cloud from filename PLY file """
    plydata = PlyData.read(filename)
    data = plydata['vertex'].data
    vertex = np.array([[x, y, z, r, g, b] for x, y, z, r, g, b in data])
    return vertex


mesh_path = "...data\scannetpp\data\0a5c013435\scans\mesh_aligned_0.05.ply"
seg_anno_path = "...data\scannetpp\data\0a5c013435\scans\segments_anno.json"
colmap_path = "...data\scannetpp\data\0a5c013435\dslr\colmap\images.txt"
colmap = []
with open(colmap_path, 'rt') as f:
    for line in f:
        col = line.strip().split()
        if col[0] == '#' or len(col) > 10: continue
        colmap.append(col)
colmap = np.array(sorted(colmap, key=lambda x: x[9])) # colume 9 is the image file name
quaternions = colmap[:, 1:5].astype(np.float32)
rots = Rotation.from_quat(quaternions).as_matrix()
trans = colmap[:, 5:8].astype(np.float32)
# all_extrins = format_poses(rots, trans)

seg_anno = json.load(open(seg_anno_path, "rt"))
vertex = read_ply(mesh_path)
print(f"Read points: {vertex.shape[0]}")
vertex = vertex[np.random.choice(vertex.shape[0], 10000, replace=False)]

segGroups = seg_anno["segGroups"]
print(f"seg groups: {len(segGroups)}")

x = np.concatenate([vertex[:, 0], trans[:, 0]], axis=0)
y = np.concatenate([vertex[:, 1], trans[:, 1]], axis=0)
z = np.concatenate([vertex[:, 2], trans[:, 2]], axis=0)
c = vertex[:, 3:] / 255
c = np.concatenate([c, np.zeros([len(trans), 3])], axis=0)
c[-len(trans):, :] = [0, 1, 0]

fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
ax.scatter(x, y, z, color=c, s=1)

for seg in segGroups:
    if seg["label"] != "wall":
        continue
    obb = seg["obb"]
    centroid = obb["centroid"]
    axesLengths = np.array(obb["axesLengths"])
    bbox_min = np.array(obb["min"])
    bbox_max = np.array(obb["max"])

    dim_x, dim_y, dim_z = np.split(bbox_max - bbox_min, 3, -1)
    # dim_x, dim_y, dim_z = np.split(axesLengths, 3, -1)
    dim_zeros = np.zeros_like(dim_x)
    corners = np.stack([
        bbox_min,
        bbox_min + np.concatenate([dim_x, dim_zeros, dim_zeros], axis=-1),
        bbox_min + np.concatenate([dim_zeros, dim_y, dim_zeros], axis=-1),
        bbox_min + np.concatenate([dim_zeros, dim_zeros, dim_z], axis=-1),
        bbox_max,
        bbox_max - np.concatenate([dim_x, dim_zeros, dim_zeros], axis=-1),
        bbox_max - np.concatenate([dim_zeros, dim_y, dim_zeros], axis=-1),
        bbox_max - np.concatenate([dim_zeros, dim_zeros, dim_z], axis=-1),
    ], axis=0)
    vertices = [corners[0], corners[1], corners[7], corners[2],
                corners[3], corners[6], corners[4], corners[5]]
    edges = [(0, 1), (1, 2), (2, 3), (3, 0),
            (4, 5), (5, 6), (6, 7), (7, 4),
            (0, 4), (1, 5), (2, 6), (3, 7)]
    for edge in edges:
        ax.plot3D(*zip(*[vertices[point] for point in edge]), color='b')

ax.set_xlabel('Width')
ax.set_ylabel('Height')
ax.set_zlabel('Depth')
plt.show()
```

**Comment by OrangeSodahub** [comment]
An obvious observation is, the scene (points) seems locates in positive space, however the tvec of camera has negative values, it's definitly wrong. Is there any addition transformation lost?

**Comment by OrangeSodahub** [comment]
@cy94 @liu115 I just checked all scenes, they are all in positive space, didn't you modify the camera poses after aligned the meshes (https://github.com/scannetpp/scannetpp/issues/25#issuecomment-1800035019)? I guess so.

**Comment by liu115** [comment]
From the code you sent, I believe that you do it transform wrongly. If you want to plot the camera position, you should use the translation of camera-to-world matrix but you are using world-to-camera instead. You could use this example code the visualize the camera pose with the mesh with open3d.


```python
"""
Visualize the camera poses + mesh using open3d
"""

import argparse
import os
import sys
from pathlib import Path

import imageio
import numpy as np
import open3d as o3d
from tqdm import tqdm

from common.utils.colmap import read_model
from common.scene_release import ScannetppScene_Release


def parser_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--data_root", type=str)
    parser.add_argument("--scene_id", type=str)
    return parser.parse_args()


def main(args):
    scene = ScannetppScene_Release(args.scene_id, data_root=args.data_root)
    mesh_path = scene.scan_mesh_path

    colmap_dir = scene.iphone_colmap_dir
    cameras, images, points3D = read_model(colmap_dir, ".txt")

    vis = o3d.visualization.Visualizer()
    vis.create_window()

    mesh = o3d.io.read_triangle_mesh(str(mesh_path))
    vis.add_geometry(mesh)

    for image_idx, image in images.items():
        world_to_camera = image.world_to_camera
        camera_to_world = np.linalg.inv(world_to_camera)
        axis = o3d.geometry.TriangleMesh.create_coordinate_frame(size=0.2)
        axis.transform(camera_to_world)
        vis.add_geometry(axis)

    vis.poll_events()
    vis.update_renderer()
    vis.run()
    vis.destroy_window()


if __name__ == "__main__":
    args = parser_args()
    main(args)
```

![image](https://github.com/scannetpp/scannetpp/assets/8998128/4846f918-c027-4710-ac27-87ec854c6a61)

**Comment by OrangeSodahub** [comment]
Thanks! I eventually found the mistake. After I add the `inv` step, it looks good. But I still think it's a little wierd. You claimed the qvec and tvec in images.txt are the extrins of camera, isn't the camera's position in the world coordinates contained in the camera's extrinsic matrix (world to camera)??? Using camera-to-world shouldn't be correct. And what I used is correct in Scannet.

Besides, after I add the `inv`, namely I used the camera-to-world matrix, in this case I successed in transforming bbox to image, and that's "camera-to-world"? It should be world-to-camera.
